### PR TITLE
File refactor

### PR DESCRIPTION
**Pull request by DylanTinianov**
Closes: #72 

## Description

This refactor abstracts the implementation of transactions, scripts, and contracts into cadence files.
What changed:
- Account models are created when queried for rather than stored in the database
- Contracts are no longer tied to accounts, and are instead executed as stand alone deployments to a specified address
- API Schema was updated to support creating contract templates and deployments
- Storage now stores all template types as Cadence Files
- Playground tests were refactored into e2eTests, and tests were added for contract templates and deployments
- Merged in projectList endpoint as well

## Work to be done later
The next step is to refactor the schema into only handling Cadence Files rather than contract, script, and transaction templates separately #100. However, this will require more frontend changes.

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/onflow/flow-playground/blob/master/CONTRIBUTING.md))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-playground/blob/master/CONTRIBUTING.md).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: adapter/models.go [hunk 1]
```diff
@@ -87,6 +87,46 @@ func ScriptToAPI(script *model.ScriptExecution) *model.ScriptExecution {
...
+}
+
+func ContractFromAPI(contract model.NewContractDeployment) model.NewContractDeployment {
+	contract.Address = AddressFromAPI(contract.Address)
+	contract.Script = ContentAddressFromAPI(contract.Script)
+	return contract
+}
+
+// TODO: Remove accounts?
+/*
```
Comment (sideninja) [inline adapter/models.go:129]: this should be removed

File: adapter/state.go [hunk 2]
```diff
@@ -18,8 +18,8 @@
 
 package adapter
 
-import "encoding/json"
-
+// TODO: Do we still need these?
+/*
 type accountState struct {
```
Comment (sideninja) [inline adapter/state.go:23]: Yes, we need to keep this, accounts will still have storages in the v2.

File: blockchain/emulator.go [hunk 3]
```diff
@@ -57,7 +57,7 @@ type blockchain interface {
 	getAccount(address flowsdk.Address) (*flowsdk.Account, *emu.AccountStorage, error)
 
 	// deployContract deploys a contract on the provided address and returns transaction and result.
-	deployContract(address flowsdk.Address, script string) (*types.TransactionResult, *flowsdk.Transaction, error)
+	deployContract(address flowsdk.Address, script string, contractName string) (*types.TransactionResult, *flowsdk.Transaction, error)
```
Comment (sideninja) [inline blockchain/emulator.go:60]: why do we need to add this value in the API?

File: blockchain/projects.go [hunk 4]
```diff
@@ -187,92 +157,104 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
 
-	return p.getAccount(em, projectID, address)
-}
+		_, err = p.Reset(&proj)
+		if err != nil {
+			return nil, err
+		}
+
+		// Reload emulator
+		em, err = p.load(projectID)
```
Comment (sideninja) [inline blockchain/projects.go:224]: I don't think this is needed, reset will clear the emulator from cache, and then first time it's tried to be accessed it will be loaded.

File: blockchain/projects.go [hunk 5]
```diff
@@ -187,92 +157,104 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
-		return nil, result.Error
-	}
 
-	exe := model.TransactionExecutionFromFlow(projectID, result, tx)
-	err = p.store.InsertTransactionExecution(exe)
+	flowAccount, _, err := em.getAccount(address.ToFlowAddress())
 	if err != nil {
 		return nil, err
 	}
+	if _, ok := flowAccount.Contracts[*contractName]; ok {
```
Comment (sideninja) [inline blockchain/projects.go:210]: We should handle this differently going forward. I say we don't reset the state anymore. If a user wants to redeploy the same contract that's already on an account we just remove that contract and deploy it again, we don't clear all other accounts and the state.

File: blockchain/emulator.go [hunk 6]
```diff
@@ -278,3 +274,11 @@ func parseContractName(code string) (string, error) {
 
 	return "", fmt.Errorf("unable to determine contract name")
 }
+
+func GetContractName(script string) (*string, error) {
```
Comment (sideninja) [inline blockchain/emulator.go:278]: why do you return string pointer? better to just return string

File: blockchain/projects.go [hunk 7]
```diff
@@ -134,50 +133,21 @@ func (p *Projects) ExecuteScript(execution model.NewScriptExecution) (*model.Scr
...
 	defer p.mutex.remove(projectID).Unlock()
-	em, err := p.load(projectID)
-	if err != nil {
-		return nil, err
-	}
-
-	return p.getAccount(em, projectID, address)
-}
-
-func (p *Projects) GetAccounts(projectID uuid.UUID, addresses []model.Address) ([]*model.Account, error) {
```
Comment (sideninja) [inline blockchain/projects.go:149]: Accounts shouldn't be removed completely, they should be kept to include `storage`, `address` and optionally `contracts`

File: blockchain/projects.go [hunk 8]
```diff
@@ -134,50 +133,21 @@ func (p *Projects) ExecuteScript(execution model.NewScriptExecution) (*model.Scr
...
-	}
-
-	return accounts, nil
+	return p.resetAccounts(projectID)
 }
 
-func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account, error) {
-	p.mutex.load(projectID).Lock()
-	defer p.mutex.remove(projectID).Unlock()
+func (p *Projects) resetAccounts(projectID uuid.UUID) (*int, error) {
```
Comment (sideninja) [inline blockchain/projects.go:143]: I don't think the name of the function reflects what's being done, this creates new accounts, but there's no reset of accounts. I think it's better if it's not renamed.

File: blockchain/projects.go [hunk 9]
```diff
@@ -187,92 +157,104 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
-		account.ID = uuid.New()
-		accounts[i] = account
 	}
 
-	return accounts, nil
+	return &p.accountsNumber, nil
 }
 
 // CreateAccount creates a new account and return the account model as well as record the execution.
-func (p *Projects) CreateAccount(projectID uuid.UUID) (*model.Account, error) {
```
Comment (sideninja) [inline blockchain/projects.go:201]: Probably need to change it back to account (won't keep commenting about the accounts)

File: blockchain/projects.go [hunk 10]
```diff
@@ -187,92 +157,104 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
+	if result.Error != nil {
+		return nil, result.Error
+	}
 
-	jsonStorage, err := json.Marshal(store)
+	exe := model.TransactionExecutionFromFlow(projectID, result, tx)
+	exe.File.Title = *contractName
+
+	deploy := model.ContractDeploymentFromFlow(projectID, result, tx)
+	deploy.File.Title = *contractName
```
Comment (sideninja) [inline blockchain/projects.go:242]: you could add it as an argument to the above factory

File: README.md [hunk 11]
```diff
@@ -11,7 +11,7 @@ make generate
 ## Testing
 
 ```shell script
-make test
+make e2eTest
```
Comment (sideninja) [inline README.md:14]: I would leave this `make test` since the test should include all the tests (including e2e). But if you want to just run e2e you can make another command `make e2eTest`

File: adapter/models.go [hunk 12]
```diff
@@ -87,12 +87,47 @@ func ScriptToAPI(script *model.ScriptExecution) *model.ScriptExecution {
...
+
+func ContractToAPI(contract *model.ContractDeployment) *model.ContractDeployment {
+	contract.Address = addressToAPI(contract.Address)
+	contract.Script = contentAddressToAPI(contract.Script)
+
+	for i, e := range contract.Errors {
+		contract.Errors[i].Message = contentAddressToAPI(e.Message)
+	}
+
+	for i, a := range contract.Events {
```
Comment (sideninja) [inline adapter/models.go:105]: I would extract this into a method called `EventsToAPI`.

File: blockchain/cache.go [hunk 13]
```diff
@@ -19,17 +19,18 @@
...
 	"github.com/google/uuid"
 )
 
 // newEmulatorCache returns a new instance of cache with provided capacity.
 func newEmulatorCache(capacity int) *emulatorCache {
 	c := lru.New(capacity)
-	c.OnEvicted = func(key lru.Key, value interface{}) {
-		fmt.Printf("Cache evicted emulator for project: %s - (%v)\n", key.(uuid.UUID).String(), key)
-	}
+	/*
```
Comment (sideninja) [inline blockchain/cache.go:29]: just remove it, don't leave commented-out code.

File: blockchain/projects.go [hunk 14]
```diff
@@ -55,21 +55,21 @@ type Projects struct {
 	accountsNumber int
 }
 
-// Reset the blockchain state.
-func (p *Projects) Reset(project *model.Project) ([]*model.Account, error) {
+// Reset the blockchain state and return the new number of accounts
+func (p *Projects) Reset(project *model.Project) (*int, error) {
```
Comment (sideninja) [inline blockchain/projects.go:59]: Why is int pointer returned? Wouldn't be better to just return new project model?

File: blockchain/projects.go [hunk 15]
```diff
@@ -187,92 +158,102 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
+		// TODO: change logic to just re-deploy this contract instead of resetting the state
+		var proj model.Project
+		err := p.store.GetProject(projectID, &proj)
+		if err != nil {
+			return nil, err
+		}
 
-	return p.getAccount(em, projectID, address)
-}
+		_, err = p.Reset(&proj)
```
Comment (sideninja) [inline blockchain/projects.go:220]: The reset method should be refactored to accept the project ID and then internally get the project.

File: blockchain/projects.go [hunk 16]
```diff
@@ -187,92 +158,102 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
+		if err != nil {
+			return nil, err
+		}
 
-func (p *Projects) getAccount(em blockchain, projectID uuid.UUID, address model.Address) (*model.Account, error) {
-	flowAccount, store, err := em.getAccount(address.ToFlowAddress())
+		// Reload emulator
+		em, err = p.load(projectID)
```
Comment (sideninja) [inline blockchain/projects.go:226]: This should be called within the reset function because otherwise resetting without calling this (which is up to the developer to not forget) can bring the project to an invalid state.

File: blockchain/projects.go [hunk 17]
```diff
@@ -369,3 +350,54 @@ func (p *Projects) filterMissingExecutions(
 
 	return executions, nil
 }
+
+// GetAccount by the address along with its storage information.
+func (p *Projects) GetAccount(projectID uuid.UUID, address model.Address) (*model.Account, error) {
+	p.mutex.load(projectID).Lock()
```
Comment (sideninja) [inline blockchain/projects.go:356]: This should be read lock only -> `RLock()`

File: blockchain/projects.go [hunk 18]
```diff
@@ -369,3 +350,54 @@ func (p *Projects) filterMissingExecutions(
...
+		}
+
+		accounts[i] = account
+	}
+
+	return accounts, nil
+}
+
+func (p *Projects) getAccount(em blockchain, projectID uuid.UUID, address model.Address) (*model.Account, error) {
+	flowAccount, store, err := em.getAccount(address.ToFlowAddress())
```
Comment (sideninja) [inline blockchain/projects.go:388]: Loading emulator should be done here `em, err := p.load(projectID)` not in the wrapper function.

File: blockchain/projects_test.go [hunk 19]
```diff
@@ -112,105 +111,7 @@ func Benchmark_LoadEmulator(b *testing.B) {
 	})
 }
 
-func Benchmark_GetAccounts(b *testing.B) {
```
Comment (sideninja) [inline blockchain/projects_test.go:115]: Why was this removed?

File: blockchain/projects_test.go [hunk 20]
```diff
@@ -620,16 +504,16 @@ func Test_DeployContract(t *testing.T) {
...
+		deploy1, err := projects.DeployContract(proj.ID, model.NewAddressFromString("0x05"), scriptA)
 		require.NoError(t, err)
-		assert.Equal(t, accA.DeployedCode, scriptA)
+		assert.Equal(t, deploy1.Title, "HelloWorldA")
 
-		accB, err := projects.DeployContract(proj.ID, accounts[1].Address, scriptB)
+		deploy2, err := projects.DeployContract(proj.ID, model.NewAddressFromString("0x06"), scriptB)
 		require.NoError(t, err)
-		assert.Equal(t, accB.DeployedCode, scriptB)
+		assert.Equal(t, deploy2.Title, "HelloWorldB")
```
Comment (sideninja) [inline blockchain/projects_test.go:516]: order of all asserts should be swapped: it is `expected value, actual value`. This is really annoying order to me as well, but it's good we stick to the correct API otherwise output is confusing.

File: blockchain/projects_test.go [hunk 21]
```diff
@@ -620,16 +504,16 @@ func Test_DeployContract(t *testing.T) {
...
 		assert.NoError(t, err)
 
-		accA, err := projects.DeployContract(proj.ID, accounts[0].Address, scriptA)
+		deploy1, err := projects.DeployContract(proj.ID, model.NewAddressFromString("0x05"), scriptA)
 		require.NoError(t, err)
-		assert.Equal(t, accA.DeployedCode, scriptA)
+		assert.Equal(t, deploy1.Title, "HelloWorldA")
 
-		accB, err := projects.DeployContract(proj.ID, accounts[1].Address, scriptB)
+		deploy2, err := projects.DeployContract(proj.ID, model.NewAddressFromString("0x06"), scriptB)
```
Comment (sideninja) [inline blockchain/projects_test.go:514]: Why is the account address no longer referenced but hardcoded by value? If this can be avoided it's better since the tests are more robust if changing values.

File: controller/accounts.go [hunk 22]
```diff
@@ -41,103 +41,43 @@ func NewAccounts(
...
-func (a *Accounts) GetByID(ID uuid.UUID, projectID uuid.UUID) (*model.Account, error) {
-	var acc model.Account
-
-	err := a.store.GetAccount(ID, projectID, &acc)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to get account")
-	}
-
-	account, err := a.blockchain.GetAccount(projectID, acc.Address)
+func (a *Accounts) GetByAddress(address model.Address, projectID uuid.UUID) (*model.Account, error) {
```
Comment (sideninja) [inline controller/accounts.go:44]: 👍

File: controller/accounts.go [hunk 23]
```diff
@@ -41,103 +41,43 @@ func NewAccounts(
...
 
 func (a *Accounts) AllForProjectID(projectID uuid.UUID) ([]*model.Account, error) {
-	var accounts []*model.Account
-
-	err := a.store.GetAccountsForProject(projectID, &accounts)
+	var proj model.Project
+	err := a.store.GetProject(projectID, &proj)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get accounts")
+		return nil, err
```
Comment (sideninja) [inline controller/accounts.go:56]: it's good to wrap the error with a more understandable message: `failed to get all project accounts`

File: controller/accounts.go [hunk 24]
```diff
@@ -41,103 +41,43 @@ func NewAccounts(
...
-	flowAccount, err = a.blockchain.DeployContract(input.ProjectID, dbAccount.Address, *input.DeployedCode)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to deploy account code")
-	}
-
-	return flowAccount.
-		MergeFromStore(&dbAccount).
-		Export(), nil
+// getAddressForAccountIndex calculates the address based on the offset from the initial account address
+func getAddressForAccountIndex(index int) model.Address {
```
Comment (sideninja) [inline controller/accounts.go:78]: This would be better to be part of the model.Address function like `NewAddressFromIndex`

File: controller/files.go [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+func (f *Files) DeleteFile(scriptID, projectID uuid.UUID) error {
+	err := f.store.DeleteCadenceFile(scriptID, projectID)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete cadence file")
+	}
+
+	return nil
+}
+
+func (f *Files) CreateScriptExecution(input model.NewScriptExecution) (*model.ScriptExecution, error) {
```
Comment (sideninja) [inline controller/files.go:80]: I feel this should be moved back to the script controller. The same is true for transactions. And DeployContract could be in a contracts controller. I think that because transactions might have more methods like getting the result of transaction and it doesn't really have to do anything with the files.

File: controller/projects.go [hunk 26]
```diff
@@ -49,60 +48,69 @@ func NewProjects(
...
+		Version:          p.version,
+		UserID:           user.ID,
 	}
 
-	ttpls := make([]*model.TransactionTemplate, len(input.TransactionTemplates))
-	for i, tpl := range input.TransactionTemplates {
-		ttpls[i] = &model.TransactionTemplate{
+	files := make(
+		[]*model.File,
+		len(input.TransactionTemplates)+len(input.ScriptTemplates)+len(input.ContractTemplates),
```
Comment (sideninja) [inline controller/projects.go:67]: In case that there's this complexity in length, is just not worth it defining the length, just define it as 0 and then append:
```suggestion
		files := make([]*model.File,0)
		files = append(files, &model.File{ ... })
```
It has insignificant impact on performance but improves readability and that's way more important here.

File: controller/projects.go [hunk 27]
```diff
@@ -49,60 +48,69 @@ func NewProjects(
...
+		return nil, errors.Wrap(err, "failed to create project")
 	}
 
-	err = p.store.InsertAccounts(accounts)
+	_, err = p.blockchain.CreateInitialAccounts(proj.ID)
 	if err != nil {
-		sentry.CaptureException(err)
 		return nil, err
 	}
+	//proj.NumberOfAccounts = *numAccounts
```
Comment (sideninja) [inline controller/projects.go:113]: remove comment if not needed

File: controller/projects.go [hunk 28]
```diff
@@ -137,6 +145,6 @@ func (p *Projects) UpdateVersion(id uuid.UUID, version *semver.Version) error {
 }
 
 // Reset is not used in the API but for migration
-func (p *Projects) Reset(proj *model.Project) ([]*model.Account, error) {
+func (p *Projects) Reset(proj *model.Project) (*int, error) {
```
Comment (sideninja) [inline controller/projects.go:148]: I feel this should be refactored as mentioned before, it should just take the project ID and return the project model.

File: controller/scripts.go [hunk 29]
```diff
@@ -18,6 +18,7 @@
 
 package controller
 
+/* TODO Delete this file
```
Comment (sideninja) [inline controller/scripts.go:21]: I think it's better to keep it and keep the execute transaction command on it.

File: controller/transactions.go [hunk 30]
```diff
@@ -99,6 +102,7 @@ func (t *Transactions) AllTemplatesForProjectID(ID uuid.UUID) ([]*model.Transact
 }
 
 func (t *Transactions) AllExecutionsForProjectID(ID uuid.UUID) ([]*model.TransactionExecution, error) {
```
Comment (sideninja) [inline controller/transactions.go:104]: Don't keep commented out code if not needed. But I feel this should at least include get tx execution

File: e2eTest/account_test.go [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+
+import (
+	"encoding/json"
+	"github.com/dapperlabs/flow-playground-api/e2eTest/client"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestAccountDeployedContracts(t *testing.T) {
```
Comment (sideninja) [inline e2eTest/account_test.go:29]: Very nice refactor of tests

File: e2eTest/client/client_test.go [hunk 32]
```diff
@@ -20,14 +20,13 @@ package client_test
 
 import (
 	"encoding/json"
+	client2 "github.com/dapperlabs/flow-playground-api/e2eTest/client"
```
Comment (sideninja) [inline e2eTest/client/client_test.go:23]: alias can be removed

File: e2eTest/constants.go [hunk 33]
```diff
@@ -0,0 +1,554 @@
+/*
+ * Flow Playground
```
Comment (sideninja) [inline e2eTest/constants.go:2]: Nice!

File: migrate/migrate.go [hunk 34]
```diff
@@ -18,11 +18,11 @@
 
 package migrate
 
-import (
-	"fmt"
+// TODO: Remove old migrators and create migrator from stable playground to playground v2
```
Comment (sideninja) [inline migrate/migrate.go:21]: I suggest either implementing this as part of this PR or making it a separate PR, but in any case don't leave commented out code.

File: blockchain/projects.go (hunk 14 above)
Comment (DylanTinianov) [reply blockchain/projects.go:59]: We don't need to return new project models since they're only created when the client actually queries for them. I should change this to an int instead of int pointer though.

File: blockchain/projects_test.go (hunk 21 above)
Comment (DylanTinianov) [reply blockchain/projects_test.go:514]: I'll change this back. It happened previously because I initially removed the accounts and thus didn't have an account to reference.

File: migrate/migrate.go (hunk 34 above)
Comment (DylanTinianov) [reply migrate/migrate.go:21]: I started another PR to create the migrator, it's planned for the next sprint.

File: README.md (hunk 11 above)
Comment (DylanTinianov) [reply README.md:14]: Looks like that got changed automatically in the readme when I did the rename, but not actually in the makefile anyways haha. Changed this back :).

File: adapter/models.go (hunk 12 above)
Comment (DylanTinianov) [reply adapter/models.go:105]: Done

File: blockchain/cache.go (hunk 13 above)
Comment (DylanTinianov) [reply blockchain/cache.go:29]: Done

File: blockchain/projects.go (hunk 15 above)
Comment (DylanTinianov) [reply blockchain/projects.go]: Done

File: blockchain/projects.go (hunk 16 above)
Comment (DylanTinianov) [reply blockchain/projects.go]: Done

File: blockchain/projects.go (hunk 17 above)
Comment (DylanTinianov) [reply blockchain/projects.go]: Done

File: blockchain/projects.go [hunk 35]
```diff
@@ -369,3 +350,54 @@ func (p *Projects) filterMissingExecutions(
...
+	em, err := p.load(projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.getAccount(em, projectID, address)
+}
+
+func (p *Projects) GetAccounts(projectID uuid.UUID, addresses []model.Address) ([]*model.Account, error) {
+	p.mutex.load(projectID).Lock()
```
[the remark this answers is not in the document]
Comment (DylanTinianov) [reply blockchain/projects.go]: Done

File: blockchain/projects.go (hunk 18 above)
Comment (DylanTinianov) [reply blockchain/projects.go:394]: Done

File: blockchain/projects_test.go (hunk 19 above)
Comment (DylanTinianov) [reply blockchain/projects_test.go:115]: Removed it when I removed accounts. Will add back.

File: controller/accounts.go (hunk 24 above)
Comment (DylanTinianov) [reply controller/accounts.go:78]: Done

File: controller/files.go (hunk 25 above)
Comment (DylanTinianov) [reply controller/files.go:80]: My thought was that contract deployments, script and transaction executions all occur based on the file / file type. I think it's more simple if we don't even have separate controllers for contracts, scripts, transactions and just have the file controller which deals with files based on their type.

CreateTransactionExecution already returns the execution which contains the result?

File: controller/accounts.go (hunk 23 above)
Comment (DylanTinianov) [reply controller/accounts.go:55]: Done

File: e2eTest/client/client_test.go (hunk 32 above)
Comment (DylanTinianov) [reply e2eTest/client/client_test.go:23]: Fixed

File: blockchain/projects_test.go [hunk 36]
```diff
@@ -569,15 +457,11 @@ func Test_DeployContract(t *testing.T) {
...
-		require.Len(t, accounts, numAccounts)
-
-		account, err := projects.DeployContract(proj.ID, accounts[0].Address, script)
+		require.Equal(t, *accounts, numAccounts)
 
+		deployment, err := projects.DeployContract(proj.ID, model.NewAddressFromString("0x05"), script)
 		require.NoError(t, err)
-		assert.Equal(t, []string{"HelloWorld"}, account.DeployedContracts)
-		assert.Equal(t, script, account.DeployedCode)
-		assert.True(t, strings.Contains(account.State, "HelloWorld"))
```
[the remark this answers is not in the document]
Comment (DylanTinianov) [reply blockchain/projects_test.go:579]: Removed this when I removed accounts previously.

File: controller/transactions.go (hunk 30 above)
Comment (DylanTinianov) [reply controller/transactions.go]: Removed the file for now

File: blockchain/projects_test.go [hunk 37]
```diff
@@ -476,12 +358,17 @@ func Test_TransactionExecution(t *testing.T) {
 				pub init() { self.A = "HelloWorldA" }
 			}`
 
-		accounts, err := projects.CreateInitialAccounts(proj.ID)
+		_, err := projects.CreateInitialAccounts(proj.ID)
 		assert.NoError(t, err)
 
-		accA, err := projects.DeployContract(proj.ID, accounts[0].Address, scriptA)
+		// Note: 0x05 is the first address (translation occurs at the API level)
+		deployment, err := projects.DeployContract(proj.ID, model.NewAddressFromString("0x05"), scriptA)
```
[the remark this answers is not in the document]
Comment (DylanTinianov) [reply blockchain/projects_test.go]: Done

File: blockchain/projects_test.go (hunk 20 above)
Comment (DylanTinianov) [reply blockchain/projects_test.go]: Fixed (at least everywhere I could see at first glance)

File: blockchain/projects_test.go (hunk 21 above)
Comment (DylanTinianov) [reply blockchain/projects_test.go]: Done

File: controller/projects.go (hunk 28 above)
Comment (DylanTinianov) [reply controller/projects.go]: Changed back to return new account models since I added accounts back anyways. I think it makes sense.

File: controller/scripts.go (hunk 29 above)
Comment (DylanTinianov) [reply controller/scripts.go]: I think it's more simple to just have the file controller that checks the file type and executes as a script, transaction, or contract based on that. Can change if this really bothers you though.

File: blockchain/projects.go [hunk 38]
```diff
@@ -55,20 +55,34 @@ type Projects struct {
...
 		return nil, err
 	}
 
-	accounts, err := p.CreateInitialAccounts(project.ID)
+	accounts, err := p.createInitialAccounts(projectID)
 	if err != nil {
 		return nil, err
 	}
 
+	// Reload emulator
```
Comment (sideninja) [inline blockchain/projects.go:78]: Why is this needed?

File: blockchain/projects.go [hunk 39]
```diff
@@ -187,11 +172,15 @@ func (p *Projects) CreateInitialAccounts(projectID uuid.UUID) ([]*model.Account,
...
 			return nil, err
 		}
+	}
 
-		account := model.AccountFromFlow(flowAccount, projectID)
-		account.Index = i
-		account.ID = uuid.New()
-		accounts[i] = account
+	var accounts []*model.Account
+	for i := 0; i < p.accountsNumber; i++ {
```
Comment (sideninja) [inline blockchain/projects.go:178]: couldn't you just collect the accounts from `createAccount` method?

File: blockchain/projects.go (hunk 38 above)
Comment (DylanTinianov) [reply blockchain/projects.go:78]: I think it is required as Reset creates new initial accounts on a new emulator to reset the state and the current emulator being used would be stale if not reloaded (for example in deploy contract). It probably won't be needed when we change contract deployment logic but I'll create another ticket for that.

**Comment by jasamba** [comment]
@DylanTinianov the PR is growing out of proportion should we consider breaking it into several smaller PRs?

File: blockchain/projects.go (hunk 39 above)
Comment (DylanTinianov) [reply blockchain/projects.go]: We still need to use p.getAccount() since there is logic there to obtain the project state from memory as well. However, nice catch that there's no need for a second loop!

**Comment by DylanTinianov** [comment]
> @DylanTinianov the PR is growing out of proportion should we consider breaking it into several smaller PRs?

@jasamba  In this case it was expected to be a large PR, as the way contract deployments were being done needed to be ripped out and fully re-implemented. It would have been difficult to break this PR into smaller components while keeping the build and tests from breaking. 

On the bright side, this PR covers the bulk of what's needed on the backend for Playground v2!

File: blockchain/projects.go [hunk 40]
```diff
@@ -369,3 +391,26 @@ func (p *Projects) filterMissingExecutions(
 
 	return executions, nil
 }
+
```
Comment (sideninja) [inline blockchain/projects.go:394]: this get account function would be better to be grouped above with the exported func

File: controller/projects.go [hunk 41]
```diff
@@ -117,6 +161,26 @@ func (p *Projects) Get(id uuid.UUID) (*model.Project, error) {
 	return &proj, nil
 }
 
+func (p *Projects) GetProjectListForUser(userID uuid.UUID, auth *auth.Authenticator, ctx context.Context) (*model.ProjectList, error) {
```
Comment (sideninja) [inline controller/projects.go:164]: I feel in the future we should move this to the user resource, so you retrieve user and all related entities like projects, tutorial completions etc. Just as a note.

File: blockchain/projects.go [hunk 42]
```diff
@@ -55,20 +55,34 @@ type Projects struct {
...
 	}
 
-	accounts, err := p.CreateInitialAccounts(project.ID)
+	accounts, err := p.createInitialAccounts(projectID)
 	if err != nil {
 		return nil, err
 	}
 
+	// Reload emulator
+	if em != nil {
```
Comment (sideninja) [inline blockchain/projects.go:79]: Please add a todo here with the link to the issue that this needs to be refactored.

File: storage/sql.go [hunk 43]
```diff
@@ -326,68 +342,52 @@ func (s *SQL) InsertTransactionExecution(exe *model.TransactionExecution) error
...
-	err := s.db.Model(&model.ScriptTemplate{}).
-		Where("project_id", tpl.ProjectID).
-		Count(&count).Error
-	if err != nil {
-		return err
-	}
-
-	tpl.Index = int(count)
-	return s.db.Create(tpl).Error
+func (s *SQL) GetContractDeploymentsForAddress(projectID uuid.UUID, address model.Address, deployments *[]*model.ContractDeployment) error {
```
Comment (sideninja) [inline storage/sql.go:360]: where is this used?

File: storage/storage.go [hunk 44]
```diff
@@ -29,40 +29,35 @@ import (
...
-	GetAccount(id, pID uuid.UUID, acc *model.Account) error
-	GetAccountsForProject(projectID uuid.UUID, accs *[]*model.Account) error
-	DeleteAccount(id, pID uuid.UUID) error
-	UpdateAccount(input model.UpdateAccount, acc *model.Account) error
+	InsertCadenceFile(file *model.File) error
+	UpdateCadenceFile(input model.UpdateFile, file *model.File) error
+	DeleteCadenceFile(id uuid.UUID, pID uuid.UUID) error
+	GetFile(id uuid.UUID, pID uuid.UUID, file *model.File) error
+	GetFilesForProject(projectID uuid.UUID, files *[]*model.File, fileType model.FileType) error
+	GetAllFilesForProject(projectID uuid.UUID, files *[]*model.File) error
```
Comment (sideninja) [inline storage/storage.go:51]: I think this should be consistently named, either `InsertCadenceFile` and `GetCadenceFile` or just `InsertFile` and `GetFile`.

File: blockchain/projects_test.go [hunk 45]
```diff
@@ -512,8 +402,9 @@ func Test_TransactionExecution(t *testing.T) {
 
 }
 
+// TODO: Add account creation to playground v2
```
Comment (sideninja) [inline blockchain/projects_test.go:405]: It's better to avoid leaving commented-out code, I feel this is something that can be deleted and we will implement tests once we have implementation.

**Review by sideninja — APPROVED** [review]
Awesome work!!!! This is a huge refactor and a lot of great stuff 🎉 
I left some minor comments but looking good.

File: storage/sql.go (hunk 43 above)
Comment (DylanTinianov) [reply storage/sql.go:360]: It's not anymore. Removed.

File: blockchain/projects.go (hunk 42 above)
Comment (DylanTinianov) [reply blockchain/projects.go:79]: Done

File: controller/projects.go (hunk 41 above)
Comment (DylanTinianov) [reply controller/projects.go:152]: Yeah sounds good!

File: storage/storage.go (hunk 44 above)
Comment (DylanTinianov) [reply storage/storage.go]: Done

File: blockchain/projects_test.go (hunk 45 above)
Comment (DylanTinianov) [reply blockchain/projects_test.go]: Removed.

File: blockchain/projects.go (hunk 40 above)
Comment (DylanTinianov) [reply blockchain/projects.go:395]: Done.
